### PR TITLE
Track generated boxes in the DOM

### PR DESCRIPTION
**Pull request by SimonSapin**
Boxes on the box tree are reference-counted with `Arc`. Boxes generated by a DOM element (including through pseudo-elements) have a second strong reference kept in that element in the DOM tree.

This is one of the pieces needed to later add incremental box construction and, after the fragment tree is similarly reference-counted, incremental layout as well as layout queries from script.

### CODE REVIEW

File: victor/src/layout/dom_traversal.rs [hunk 1]
```diff
@@ -39,10 +40,47 @@ pub(super) trait TraversalHandler {
...
+}
+
+impl Drop for BoxSlot<'_> {
+    fn drop(&mut self) {
+        assert!(self.slot.is_some(), "failed to set a layout box")
+    }
+}
+
+impl Context<'_> {
+    fn layout_data(&self, element_id: NodeId) -> AtomicRefMut<LayoutDataForElement> {
```
Comment (nox) [inline victor/src/layout/dom_traversal.rs]: `layout_data_mut`

File: victor/src/layout/dom_traversal.rs [hunk 2]
```diff
@@ -39,10 +40,47 @@ pub(super) trait TraversalHandler {
...
+    }
+
+    pub fn set(mut self, box_: LayoutBox) {
+        *self.slot = Some(box_)
+    }
+}
+
+impl Drop for BoxSlot<'_> {
+    fn drop(&mut self) {
+        assert!(self.slot.is_some(), "failed to set a layout box")
```
Comment (nox) [inline victor/src/layout/dom_traversal.rs]: Won't we want to reset that to `None` for elements which became `display:contents` between two dynamic layouts?

**Review by nox — COMMENTED** [review]
Seems good.

File: victor/src/layout/dom_traversal.rs (hunk 1 above)
Comment (SimonSapin) [reply victor/src/layout/dom_traversal.rs]: Done.

File: victor/src/layout/dom_traversal.rs (hunk 2 above)
Comment (SimonSapin) [reply victor/src/layout/dom_traversal.rs]: I think this code is ok, but your comment lead to find another bug which is fixed in 1340d0679aa4f7e6000ea66a26999af21cc431d7.

**Pull request merged**